src/syscore/windows/iocp/iocp.rs

```diff
@@ -0,0 +1,352 @@
+use super::Processor;
+use crate::handle::Handle;
+use crate::submission_handler::SubmissionHandler;
+use futures::channel::oneshot;
+use futures::io::{AsyncRead, AsyncWrite};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::mem::ManuallyDrop;
+use std::net::TcpStream;
+use std::os::windows::io::{AsRawHandle, FromRawHandle, RawHandle};
+use std::ptr;
+use std::time::Duration;
+use std::{io, task};
+use std::{
+    pin::Pin,
+    sync::atomic::{AtomicUsize, Ordering},
+    task::Context,
+    task::Poll,
+};
+use winapi::um::handleapi;
+use winapi::um::handleapi::INVALID_HANDLE_VALUE;
+use winapi::um::ioapiset;
+use winapi::um::minwinbase::OVERLAPPED_ENTRY;
+use winapi::um::winbase::INFINITE;
+use winapi::um::winnt::HANDLE;
+
+pub struct SysProactor {
+    iocp_handle: WinHandle,
+    completions: HashMap<WinHandle, oneshot::Sender<usize>>,
+    // ULONG_PTR
+    completion_key: AtomicUsize,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct WinOverlappedEntry(OVERLAPPED_ENTRY);
+
+impl std::ops::Deref for WinOverlappedEntry {
+    type Target = OVERLAPPED_ENTRY;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<OVERLAPPED_ENTRY> for WinOverlappedEntry {
+    fn from(h: OVERLAPPED_ENTRY) -> Self {
+        Self(h)
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct WinHandle(HANDLE);
+
+// TODO [igni]: SUPER CURSED THIS NEEDS TO GO AWAY
```

Comment:
Actually it might not be that cursed on windows versions newer than 10. The default threadpool allows HWNDs to cross FFI bounds safely because all of the api calls are ran on the system-owned threadpool.